[Cache] Add PhpArrayAdapter to use shared memory on PHP 7.0

tgalopin opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | WIP |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Depends on https://github.com/symfony/symfony/pull/18825

This PR aims to implement a Symfony Cache adapter able to build files optimized for OPCache memory storage. Any kind of static data that can be warmed up is a candidate for OPcache storage.

This PR is limited to the implementation of the Adapter. It's usage in the annotations caching system is implemented in the PR https://github.com/symfony/symfony/pull/18533. The aim is also to use this adapter in other contexts (validator, serializer, ...).

tgalopin committed:
[Cache] Add PhpArrayAdapter to use shared memory on PHP 7.0+

stof commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,348 @@
+
+        if ('N;' === $value) {
+            $value = null;
+        } elseif (isset($value[2]) && is_string($value) && ':' === $value[1]) {
```
``is_string`` should come before ``isset($value[2])`` IMO

phansys commented on src/Symfony/Component/Cache/Adapter/NullAdapter.php:
```diff
@@ -0,0 +1,121 @@
+            yield $key => $f($key);
+        }
+    }
+}
```
Missing LF at EOF.

phansys commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,348 @@
+            }
+        }
+
+        $dump = <<<EOF
```
This should be enclosed with nowdoc instead of heredoc.

OskarStark commented on src/Symfony/Component/Cache/Adapter/NullAdapter.php:
```diff
@@ -0,0 +1,121 @@
+ */
+class NullAdapter implements AdapterInterface
+{
+    private $createCacheItem;
```
phpdoc missing

OskarStark commented on src/Symfony/Component/Cache/Adapter/NullAdapter.php:
```diff
@@ -0,0 +1,121 @@
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class NullAdapter implements AdapterInterface
```
i think this `NullAdapter` belongs to another PR which you already did, or am i wrong?

#18825

tgalopin replied:
This PR is based on #18825 which means that when #18825 will be merged I'll rebase this one on master to remove the commit. In the meantime, I use the commit for the tests.

OskarStark commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,348 @@
+    private $values;
+    private $createCacheItem;
+
+    /** @var AdapterInterface */
```
don't know the symfony standard, but i would write

```php
/**
 * @var AdapterInterface
 */
```

instead of one line

OskarStark commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,348 @@
+{
+    use LoggerAwareTrait;
+
+    private $file;
```
do we need php doc for private properties

stof replied:
not when the constructor allows IDE to infer their type and when the name is clear enough (and if the name is not clear, the preferred way is to use a better name)

OskarStark commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,348 @@
+            }
+        }
+    }
+}
```
Missing line ending

OskarStark commented on src/Symfony/Component/Cache/Tests/Adapter/NullAdapterTest.php:
```diff
@@ -0,0 +1,124 @@
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\Adapter\NullAdapter;
+
+class NullAdapterTest extends \PHPUnit_Framework_TestCase
```
this belongs to the other PR as well

tgalopin replied:
Same reason

OskarStark commented on src/Symfony/Component/Cache/Tests/Adapter/OpCacheAdapterTest.php:
```diff
@@ -0,0 +1,127 @@
+
+        return true;
+    }
+}
```
Missing line ending

OskarStark commented on src/Symfony/Component/Cache/Tests/Adapter/OpCacheAdapterWithFallbackTest.php:
```diff
@@ -0,0 +1,53 @@
+
+        return new OpCacheAdapter(self::$file, new FilesystemAdapter('opcache-fallback'));
+    }
+}
```
missing line ending

tgalopin commented:
I did the changes

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,358 @@
+EOF;
+
+        foreach ($values as $key => $value) {
+            CacheItem::validateKey($key);
```
because PHP casts numeric string keys to ints, we should cast ints to string here:
`is_int($key) ? (string) $key : $key);

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,358 @@
+            $this->initialize();
+        }
+
+        if (!isset($this->values[$key])) {
```
let's validate before that $key is a string to prevent issues with magic array index casting (same below)

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,366 @@
+        $fallbackKeys = array();
+
+        foreach ($keys as $key) {
+            if (!is_string($key)) {
```
this check should be done before calling generateItems

stof replied:
indeed. Doing it in the generator will throw the exception too late (the iterator will throw it during iteration, while it must be throw by getItems)

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,366 @@
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
```
missing is_string check

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,366 @@
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
```
missing is_string check

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php:
```diff
@@ -0,0 +1,366 @@
+        $fallbackKeys = array();
+
+        foreach ($keys as $key) {
+            if (isset($this->values[$key])) {
```
missing is_string check

nicolas-grekas commented:
I suggest renaming this adapter to Php7Adapter (ans use the OpcacheAdapter name in #18894)

tgalopin commented:
I did the changes

nicolas-grekas commented:
:+1:

nicolas-grekas commented:
Btw I looked at this umask issue and it happens that Symfony considers this as an ops issue that Symfony should not deal with. Let's remove it from this PR.

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/Php7Adapter.php:
```diff
@@ -0,0 +1,376 @@
+
+                if ('N;' === $value) {
+                    $value = null;
+                } elseif (isset($value[2]) && is_string($value) && ':' === $value[1]) {
```
is_string should be first

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/Php7Adapter.php:
```diff
@@ -0,0 +1,376 @@
+    /**
+     * @param string           $file         The PHP file were values are cached.
+     * @param AdapterInterface $fallbackPool A pool to fallback on when an item is not hit.
+     * @param int              $umask
```
IMHO we should drop this umask concern: there is no other place in the Symfony code base where we deal with this, which means that we've always considered that a wrong system umask (zero), is an ops mistake, and it's not Symfony's job to try to fix external mistakes. We've had the same reasonings about e.g. Varnish misconfigs.

tgalopin commented:
I did the changes

nicolas-grekas commented:
I renamed #18894 back to PhpFilesAdapter, and I now think this one should be renamed PhpArrayAdapter. Easier to understand, more explicit complementarity.

tgalopin commented:
I updated this

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,367 @@
+/**
+ * Adapter building static PHP files that will be cached by OPCache.
+ * This Adapter is in read-only if you use AdapterInterface methods.
+ * You can use the method "store" to build the cache file.
```
store -> warmUp

nicolas-grekas commented:
Calling for votes @symfony/deciders

fabpot commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,367 @@
+    private $fallbackPool;
+
+    /**
+     * @param string $file The PHP file were values are cached.
```
We need a description of this method usage and it should be moved after the constructor.

fabpot commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,367 @@
+     *
+     * @return CacheItemPoolInterface
+     */
+    public static function createOnPhp7($file, CacheItemPoolInterface $fallbackPool)
```
I don't like the method name. I suppose that this will also work on PHP 8, right? So including a version in a method name sounds fragile to me. But then, I don't understand the usage, so I might be wrong.

nicolas-grekas replied:
I suggest using simply `create` and ensure the doc block is clear enough about it's behavior

tgalopin commented:
I updated this

tgalopin commented:
I fixed a problem with HHVM

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,379 @@
+
+                return $item;
+            },
+            $this,
```
`null` instead of `$this`

tgalopin replied:
Well, shouldn't we remove the bind then?

nicolas-grekas replied:
the next line is the important one

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,379 @@
+/**
+ * Adapter building static PHP files that will be cached by OPCache.
+ * This Adapter is in read-only if you use AdapterInterface methods.
+ * You can use the method "warmUp" to build the cache file.
```
I suggest:
```
 * Caches items at warm up time using a static PHP array that is stored in shared memory by OPCache since PHP 5.6.
 * Warmed up items are read-only and run-time discovered items are cached using a fallback adapter.
```

fabpot commented:
@tgalopin Can you take comments into account? Looks like we are almost there :)

tgalopin commented:
I did some benchmarks to check the PHP 5.6 support in this PR.

I used three Docker containers with only PHP/HHVM installed, on the same host. I used Apache Benchmark with 1000 requests at a concurrency of 100 (`ab -n 1000 -c 100`).

The results are the following:

| Engine    	| Context              	  	| Time per request 	|
| -----------		| ------------------------		| ------------------		|
| PHP 5.6   	| Empty array            	| 0.953 ms/req     	|
| PHP 5.6   	| Array with 15 000 keys 	| 4.671 ms/req     	|
| PHP 7.0   	| Empty array            	| 0.688 ms/req     	|
| PHP 7.0   	| Array with 15 000 keys 	| 0.709 ms/req     	|
| HHVM 3.12 	| Empty array            	| 0.869 ms/req     	|
| HHVM 3.12 	| Array with 15 000 keys 	| 0.875 ms/req     	|

This seems to indicate that having an array filled with data in PHP 5.6 is significantly slower than having an empty array. Therefore, I removed the PHP 5.6 support for this PhpArrayAdapter in the `create` factory method (and I added the HHVM support).

nicolas-grekas commented:
:+1:

dunglas commented:
👍

xabbuh commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,358 @@
+                try {
+                    $value = serialize($value);
+                } catch (\Exception $e) {
+                    throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable %s value', $key, get_class($value)), 0, $e);
```
Missing closing double quote.

fabpot replied:
Also missing a dot at the end.

fabpot commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,358 @@
+                    $serialized = serialize($value);
+                    $unserialized = unserialize($serialized);
+                } catch (\Exception $e) {
+                    throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable array value', $key), 0, $e);
```
missing dot at the end (and probably on other exception messages).

xabbuh commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,358 @@
+    {
+        $this->values = array();
+
+        @unlink($this->file);
```
Don't we have to return `false` too if it was not possible to delete the file?

nicolas-grekas replied:
Let's do by using `file_exists` after unlinking when `unlink` returns `false`. Even if it opens a race condition, warming up the cache should be rare and slower than the time between the `unlink` and `file_exists`, so that in reality the race doesn't happen.

tgalopin commented:
I did the changes.

fabpot commented on src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:
```diff
@@ -0,0 +1,358 @@
+    private $fallbackPool;
+
+    /**
+     * @param string           $file         The PHP file were values are cached.
```
dots at the end of @param (@return, @throws, ... as well) should be removed

tgalopin commented:
I did the change

fabpot commented:
Thank you @tgalopin.
